src/main/scala/workflow/graph/Pipeline.scala

```diff
@@ -13,13 +13,48 @@ import scala.reflect.ClassTag
  *
  * Warning: Not thread-safe!
  *
+ * @param executor The [[GraphExecutor]] underlying the Pipeline execution.
+ * @param source The SourceId of the Pipeline
+ * @param sink The SinkId of the Pipeline
  * @tparam A type of the data this Pipeline expects as input
  * @tparam B type of the data this Pipeline outputs
  */
-trait Pipeline[A, B] {
-  private[graph] val source: SourceId
-  private[graph] val sink: SinkId
-  private[graph] def executor: GraphExecutor
+class Pipeline[A, B] private[graph] (
+  private[graph] val executor: GraphExecutor,
+  private[graph] val source: SourceId,
+  private[graph] val sink: SinkId) extends Chainable[A, B] {
+
+  private[graph] def toPipeline: Pipeline[A, B] = this
+
+  // TODO: Add docs
+  final def fit(): FittedPipeline[A, B] = {
+    val optimizedGraph = Pipeline.getOptimizer.execute(executor.graph, Map())._1
+
+    val estFittingExecutor = new GraphExecutor(optimizedGraph, optimize = false)
+    val delegatingNodes = optimizedGraph.operators.collect {
+      case (node, _: DelegatingOperator) => node
+    }
+
+    val graphWithFitEstimators = delegatingNodes.foldLeft(optimizedGraph) {
+      case (curGraph, node) => {
+        val deps = optimizedGraph.getDependencies(node)
+        val estimatorDep = deps.head
+        val transformer = estFittingExecutor.execute(estimatorDep).get.asInstanceOf[TransformerOperator]
+
+        curGraph.setOperator(node, transformer).setDependencies(node, deps.tail)
+      }
+    }
+
+    val newGraph = UnusedBranchRemovalRule.apply(graphWithFitEstimators, Map())._1
+
+    val transformerGraph = TransformerGraph(
+      newGraph.sources,
+      newGraph.sinkDependencies,
+      newGraph.operators.map(op => (op._1, op._2.asInstanceOf[TransformerOperator])),
+      newGraph.dependencies)
+
+    new FittedPipeline[A, B](transformerGraph, source, sink)
```

Comment:
`Operator` isn't guaranteed to be serializable, but `TransformerOperator` is. I'll note that in the documentation.